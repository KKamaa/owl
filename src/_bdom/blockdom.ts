import {
  Block,
  BLOCK_TYPE,
  Builder,
  Anchor,
  BlockElement,
  BlockMulti,
  BlockList,
  BlockText,
} from "./types";

// -----------------------------------------------------------------------------
//  Factories
// -----------------------------------------------------------------------------

export function text(text: string): BlockText {
  return {
    type: BLOCK_TYPE.Text,
    el: undefined,
    key: undefined,
    data: text,
    content: undefined,
  };
}

export function elem(builder: Builder, data: any[] = [], children: Block[] = []): BlockElement {
  return {
    type: BLOCK_TYPE.Element,
    el: undefined,
    key: undefined,
    data: { builder, data },
    content: children,
  };
}

export function multi(blocks: (Block | undefined)[]): BlockMulti {
  return {
    type: BLOCK_TYPE.Multi,
    el: undefined,
    key: undefined,
    data: undefined,
    content: blocks,
  };
}

export function list(
  blocks: Block[],
  isOnlyChild: boolean = false,
  hasNoComponent: boolean = false
): BlockList {
  return {
    type: BLOCK_TYPE.List,
    el: undefined,
    key: undefined,
    data: { anchor: undefined, isOnlyChild, hasNoComponent },
    content: blocks,
  };
}
// -----------------------------------------------------------------------------
//  BDom manipulation
// -----------------------------------------------------------------------------

export function mountBefore(block: Block, anchor: Anchor) {
  switch (block.type) {
    case BLOCK_TYPE.Text:
      {
        let el = document.createTextNode(block.data);
        block.el = el;
        anchor.before(el);
      }
      break;
    case BLOCK_TYPE.Element:
      {
        const data = block.data;
        const builder = new data.builder(data, block.content);
        data.builder = builder;
        const el = builder.el;
        anchor.before(el);
        block.el = el;
      }
      break;
    case BLOCK_TYPE.Multi:
      {
        const children = block.content;
        const anchors = new Array(children.length);
        for (let i = 0, l = children.length; i < l; i++) {
          let child = children[i];
          const childAnchor = document.createTextNode("");
          anchor.before(childAnchor);
          anchors![i] = childAnchor;
          if (child) {
            mountBefore(child, childAnchor);
          }
        }
        block.data = anchors;
      }
      break;
    case BLOCK_TYPE.List: {
      const children = block.content;
      const _anchor = document.createTextNode("");
      block.data.anchor = _anchor;
      anchor.before(_anchor);
      for (let i = 0, l = children.length; i < l; i++) {
        mountBefore(children[i], _anchor);
      }
    }
  }
}

function moveBefore(block: Block, anchor: Anchor | ChildNode) {
  switch (block.type) {
    case BLOCK_TYPE.Text:
      {
        anchor.before(block.el!);
      }
      break;
    case BLOCK_TYPE.Element:
      {
        anchor.before(block.el!);
      }
      break;
    case BLOCK_TYPE.Multi:
      {
        // console.warn('movebefore', block, (anchor as any).innerHTML)
        const children = block.content;
        const anchors = block.data!;
        for (let i = 0, l = children.length; i < l; i++) {
          let child = children[i];
          let _anchor = anchors[i];
          // console.warn(child!.el!.nextSibling === _anchor)
          // console.warn('c', child)
          anchor.before(_anchor);
          if (child) {
            moveBefore(child, _anchor);
          }
        }
      }
      break;
    case BLOCK_TYPE.List: {
      // const children = block.content;
      // const _anchor = document.createTextNode("");
      // block.data.anchor = _anchor;
      // anchor.before(_anchor);
      // for (let i = 0, l = children.length; i < l; i++) {
      //   mountBefore(children[i], _anchor);
      // }
    }
  }
}

export function removeBlock(block: Block) {
  switch (block.type) {
    case BLOCK_TYPE.Text:
    case BLOCK_TYPE.Element:
      block.el!.remove();
      break;
    case BLOCK_TYPE.Multi:
      {
        const children = block.content;
        const anchors = block.data;
        for (let i = 0; i < children.length; i++) {
          const child = children[i];
          if (child) {
            remove(child);
          }
          anchors![i].remove();
        }
      }
      break;
    case BLOCK_TYPE.List: {
      const { isOnlyChild, anchor } = block.data;
      if (isOnlyChild) {
        anchor!.parentElement!.textContent = "";
      } else {
        const children = block.content;
        for (let i = 0, l = children.length; i < l; i++) {
          removeBlock(children[i]);
        }
        anchor!.remove();
      }
    }
  }
}

export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  switch (block1.type) {
    case BLOCK_TYPE.Text:
      {
        let text = block2.data as any;
        if (block1.data !== text) {
          block1.data = text;
          block1.el!.textContent = text;
        }
      }
      break;
    case BLOCK_TYPE.Element:
      (block1 as any).data.builder.update((block2 as any).data, block2.content);
      break;
    case BLOCK_TYPE.Multi:
      {
        const children = block1.content;
        const newChildren = block2.content!;
        const anchors = block1.data!;
        for (let i = 0, l = children.length; i < l; i++) {
          const block = children[i];
          const newBlock = newChildren[i];
          if (block) {
            if (newBlock) {
              patch(block, newBlock);
            } else {
              children[i] = undefined;
              remove(block);
            }
          } else if (newBlock) {
            children[i] = newBlock;
            mountBefore(newBlock, anchors[i]);
          }
        }
        // block1.content = block2.content!;
      }
      break;
    case BLOCK_TYPE.List:
      {
        const oldCh = block1.content;
        const newCh: Block[] = (block2 as any).content;
        const data = block1.data;
        const _anchor = data.anchor!;

        // fast path
        if (newCh.length === 0 && data.isOnlyChild) {
          if (!data.hasNoComponent) {
            for (let i = 0; i < oldCh.length; i++) {
              beforeRemove(oldCh[i]);
            }
          }

          const parent = _anchor.parentElement!;
          _anchor.remove();
          parent.textContent = "";
          parent.appendChild(_anchor);
          block1.content = newCh;
          return;
        }

        let oldStartIdx = 0;
        let newStartIdx = 0;
        let oldStartBlock = oldCh[0];
        let newStartBlock = newCh[0];

        let oldEndIdx = oldCh.length - 1;
        let newEndIdx = newCh.length - 1;
        let oldEndBlock = oldCh[oldEndIdx];
        let newEndBlock = newCh[newEndIdx];

        let mapping: any = undefined;
        let noFullRemove = data.hasNoComponent;

        // console.warn('oldch', oldCh[0].content)
        while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
          // console.warn('iteration', oldStartIdx, oldEndIdx, newStartIdx, newEndIdx);
          // -------------------------------------------------------------------
          if (oldStartBlock === null) {
            // todo: comment this and check it is useful
            oldStartBlock = oldCh[++oldStartIdx];
          }
          // -------------------------------------------------------------------
          else if (oldEndBlock === null) {
            oldEndBlock = oldCh[--oldEndIdx];
          }
          // -------------------------------------------------------------------
          else if (newStartBlock === null) {
            // todo: understand why these 2 next elseifs are useful
            newStartBlock = newCh[++newStartIdx];
          }
          // -------------------------------------------------------------------
          else if (newEndBlock === null) {
            newEndBlock = newCh[--newEndIdx];
          }
          // -------------------------------------------------------------------
          else if (oldStartBlock.key === newStartBlock.key) {
            // console.warn('same start', oldStartBlock, newStartBlock);
            patch(oldStartBlock, newStartBlock);
            newCh[newStartIdx] = oldStartBlock;
            oldStartBlock = oldCh[++oldStartIdx];
            newStartBlock = newCh[++newStartIdx];
          }
          // -------------------------------------------------------------------
          else if (oldEndBlock.key === newEndBlock.key) {
            // console.warn('same end');
            patch(oldEndBlock, newEndBlock);
            // console.warn('oeb', oldEndBlock)
            newCh[newEndIdx] = oldEndBlock;
            // console.warn(newCh[newEndIdx])
            oldEndBlock = oldCh[--oldEndIdx];
            newEndBlock = newCh[--newEndIdx];
          }
          // -------------------------------------------------------------------
          else if (oldStartBlock.key === newEndBlock.key) {
            // bnode moved right
            patch(oldStartBlock, newEndBlock);
            const nextChild = newCh[newEndIdx + 1];
            const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
            moveBefore(oldStartBlock, anchor);
            newCh[newEndIdx] = oldStartBlock;
            oldStartBlock = oldCh[++oldStartIdx];
            newEndBlock = newCh[--newEndIdx];
          }
          // -------------------------------------------------------------------
          else if (oldEndBlock.key === newStartBlock.key) {
            // bnode moved left
            patch(oldEndBlock, newStartBlock);
            const nextChild = oldCh[oldStartIdx];
            const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
            moveBefore(oldEndBlock, anchor);
            newCh[newStartIdx] = oldEndBlock;
            oldEndBlock = oldCh[--oldEndIdx];
            newStartBlock = newCh[++newStartIdx];
          }
          // -------------------------------------------------------------------
          else {
            // console.warn('in else')
            mapping = mapping || createMapping(oldCh, oldStartIdx, oldEndIdx);
            // console.warn('mapping', mapping)
            let idxInOld = mapping[newStartBlock.key];
            if (idxInOld === undefined) {
              // new element
              mountBefore(
                newStartBlock,
                oldStartBlock.el || (firstChildNode(oldStartBlock)! as any)
              );
              // console.log(document.body.innerHTML)
            } else {
              const elmToMove = oldCh[idxInOld];
              moveBefore(elmToMove, oldStartBlock.el || (firstChildNode(oldStartBlock) as any));
              patch(elmToMove, newStartBlock);
              newCh[newStartIdx] = elmToMove;
              oldCh[idxInOld] = null as any;
            }
            newStartBlock = newCh[++newStartIdx];
          }
        }
        // ---------------------------------------------------------------------
        // console.warn('after')
        if (oldStartIdx <= oldEndIdx || newStartIdx <= newEndIdx) {
          if (oldStartIdx > oldEndIdx) {
            const nextChild = newCh[newEndIdx + 1];
            // console.warn('next', nextChild)
            const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
            for (let i = newStartIdx; i <= newEndIdx; i++) {
              mountBefore(newCh[i], anchor as any);
            }
          } else {
            for (let i = oldStartIdx; i <= oldEndIdx; i++) {
              let ch = oldCh[i];
              if (ch) {
                if (noFullRemove) {
                  remove(ch);
                } else {
                  removeBlock(ch);
                }
              }
            }
          }
        }
        block1.content = newCh;
      }
      break;
  }
}

function firstChildNode(block: Block): ChildNode | null {
  switch (block.type) {
    case BLOCK_TYPE.Multi:
      {
        const children = block.content;
        for (let i = 0, l = children.length; i < l; i++) {
          const child = children[i];
          if (child) {
            return child.el || firstChildNode(child);
          }
        }
      }
      break;
    case BLOCK_TYPE.List: {
      const first = block.content[0];
      return first ? first.el || firstChildNode(first) : null;
    }
  }
  return null;
}

export function remove(block: Block) {
  beforeRemove(block);
  removeBlock(block);
}

function createMapping(
  oldCh: any[],
  oldStartIdx: number,
  oldEndIdx: number
): { [key: string]: any } {
  let mapping: any = {};
  for (let i = oldStartIdx; i <= oldEndIdx; i++) {
    mapping[oldCh[i].key] = i;
  }
  return mapping;
}
