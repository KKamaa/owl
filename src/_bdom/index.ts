import { Block } from "./types";

export { makeBlock } from "./element";
export { list } from "./list";
export { multi } from "./multi";
export { text } from "./text";

// export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  block.ops.mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  block1.ops.patch(block1, block2);
}

export function remove(block: Block) {
  // beforeRemove(block);
  block.ops.remove(block);
}
