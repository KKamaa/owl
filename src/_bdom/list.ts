import { Anchor, Block, BlockList, Operations } from "./types";

// -----------------------------------------------------------------------------
//  List blocks
// -----------------------------------------------------------------------------

export function list(
  blocks: Block[],
  isOnlyChild: boolean = false,
  hasNoComponent: boolean = false
): BlockList {
  return {
    ops: LIST_OPS,
    el: undefined,
    key: undefined,
    data: { anchor: undefined, isOnlyChild, hasNoComponent },
    content: blocks,
  };
}

const LIST_OPS: Operations = {
  mountBefore(block: any, anchor: Anchor) {
    const children = block.content;
    const _anchor = document.createTextNode("");
    block.data.anchor = _anchor;
    anchor.before(_anchor);
    const l = children.length;
    if (l) {
      const mountBefore = children[0].ops.mountBefore;
      for (let i = 0; i < l; i++) {
        mountBefore(children[i], _anchor);
      }
    }
  },
  patch(block1: any, block2: any) {
    if (block1 === block2) {
      return;
    }
    const oldCh = block1.content;
    const newCh: Block[] = (block2 as any).content;
    if (newCh.length === 0 && oldCh.length === 0) {
      return;
    }
    const ops: Operations = oldCh[0] ? oldCh[0].ops : newCh[0].ops;
    const { firstChildNode, moveBefore, mountBefore, patch, remove: removeBlock } = ops;

    const data = block1.data;
    const _anchor = data.anchor!;

    // fast path
    if (newCh.length === 0 && data.isOnlyChild) {
      // if (!data.hasNoComponent) {
      //   for (let i = 0; i < oldCh.length; i++) {
      //     beforeRemove(oldCh[i]);
      //   }
      // }

      const parent = _anchor.parentElement!;
      _anchor.remove();
      parent.textContent = "";
      parent.appendChild(_anchor);
      block1.content = newCh;
      return;
    }

    let oldStartIdx = 0;
    let newStartIdx = 0;
    let oldStartBlock = oldCh[0];
    let newStartBlock = newCh[0];

    let oldEndIdx = oldCh.length - 1;
    let newEndIdx = newCh.length - 1;
    let oldEndBlock = oldCh[oldEndIdx];
    let newEndBlock = newCh[newEndIdx];

    let mapping: any = undefined;
    let noFullRemove = data.hasNoComponent;

    while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
      // -------------------------------------------------------------------
      if (oldStartBlock === null) {
        oldStartBlock = oldCh[++oldStartIdx];
      }
      // -------------------------------------------------------------------
      else if (oldEndBlock === null) {
        oldEndBlock = oldCh[--oldEndIdx];
      }
      // -------------------------------------------------------------------
      else if (oldStartBlock.key === newStartBlock.key) {
        patch(oldStartBlock, newStartBlock);
        newCh[newStartIdx] = oldStartBlock;
        oldStartBlock = oldCh[++oldStartIdx];
        newStartBlock = newCh[++newStartIdx];
      }
      // -------------------------------------------------------------------
      else if (oldEndBlock.key === newEndBlock.key) {
        patch(oldEndBlock, newEndBlock);
        newCh[newEndIdx] = oldEndBlock;
        oldEndBlock = oldCh[--oldEndIdx];
        newEndBlock = newCh[--newEndIdx];
      }
      // -------------------------------------------------------------------
      else if (oldStartBlock.key === newEndBlock.key) {
        // bnode moved right
        patch(oldStartBlock, newEndBlock);
        const nextChild = newCh[newEndIdx + 1];
        const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
        moveBefore(oldStartBlock, anchor);
        newCh[newEndIdx] = oldStartBlock;
        oldStartBlock = oldCh[++oldStartIdx];
        newEndBlock = newCh[--newEndIdx];
      }
      // -------------------------------------------------------------------
      else if (oldEndBlock.key === newStartBlock.key) {
        // bnode moved left
        patch(oldEndBlock, newStartBlock);
        const nextChild = oldCh[oldStartIdx];
        const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
        moveBefore(oldEndBlock, anchor);
        newCh[newStartIdx] = oldEndBlock;
        oldEndBlock = oldCh[--oldEndIdx];
        newStartBlock = newCh[++newStartIdx];
      }
      // -------------------------------------------------------------------
      else {
        mapping = mapping || createMapping(oldCh, oldStartIdx, oldEndIdx);
        let idxInOld = mapping[newStartBlock.key];
        if (idxInOld === undefined) {
          mountBefore(newStartBlock, oldStartBlock.el || (firstChildNode(oldStartBlock)! as any));
        } else {
          const elmToMove = oldCh[idxInOld];
          moveBefore(elmToMove, oldStartBlock.el || (firstChildNode(oldStartBlock) as any));
          patch(elmToMove, newStartBlock);
          newCh[newStartIdx] = elmToMove;
          oldCh[idxInOld] = null as any;
        }
        newStartBlock = newCh[++newStartIdx];
      }
    }
    // ---------------------------------------------------------------------
    if (oldStartIdx <= oldEndIdx || newStartIdx <= newEndIdx) {
      if (oldStartIdx > oldEndIdx) {
        const nextChild = newCh[newEndIdx + 1];
        const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
        for (let i = newStartIdx; i <= newEndIdx; i++) {
          mountBefore(newCh[i], anchor as any);
        }
      } else {
        for (let i = oldStartIdx; i <= oldEndIdx; i++) {
          let ch = oldCh[i];
          if (ch) {
            if (noFullRemove) {
              // remove(ch);
              removeBlock(ch);
            } else {
              removeBlock(ch);
            }
          }
        }
      }
    }
    block1.content = newCh;
  },
  moveBefore(block: any, anchor: any) {
    // todo
  },
  remove(block: any) {
    const { isOnlyChild, anchor } = block.data;
    if (isOnlyChild) {
      anchor!.parentElement!.textContent = "";
    } else {
      const children = block.content;
      const l = children.length;
      if (l) {
        const removeBlock = children[0].ops.remove;
        for (let i = 0; i < l; i++) {
          removeBlock(children[i]);
        }
      }
      anchor!.remove();
    }
  },
  firstChildNode(block: any): ChildNode | null {
    const first = block.content[0];
    return first ? first.el || first.ops.firstChildNode(first) : null;
  },
};

function createMapping(
  oldCh: any[],
  oldStartIdx: number,
  oldEndIdx: number
): { [key: string]: any } {
  let mapping: any = {};
  for (let i = oldStartIdx; i <= oldEndIdx; i++) {
    mapping[oldCh[i].key] = i;
  }
  return mapping;
}
