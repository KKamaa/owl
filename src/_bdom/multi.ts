import { Anchor, Block, BlockMulti, Operations } from "./types";

// -----------------------------------------------------------------------------
//  Multi blocks
// -----------------------------------------------------------------------------

export function multi(blocks: (Block | undefined)[]): BlockMulti {
  return {
    ops: MULTI_BLOCK_OPS,
    el: undefined,
    data: undefined,
    content: blocks,
  };
}

const MULTI_BLOCK_OPS: Operations = {
  mountBefore(block: any, anchor: Anchor) {
    const children = block.content;
    const anchors = new Array(children.length);
    for (let i = 0, l = children.length; i < l; i++) {
      let child = children[i];
      const childAnchor = document.createTextNode("");
      anchor.before(childAnchor);
      anchors![i] = childAnchor;
      if (child) {
        child.ops.mountBefore(child, childAnchor);
      }
    }
    block.data = anchors;
  },
  patch(block1: any, block2: any) {
    if (block1 === block2) {
      return;
    }
    const children = block1.content;
    const newChildren = block2.content!;
    const anchors = block1.data!;
    for (let i = 0, l = children.length; i < l; i++) {
      const block = children[i];
      const newBlock = newChildren[i];
      if (block) {
        if (newBlock) {
          block.ops.patch(block, newBlock);
        } else {
          children[i] = undefined;
          block.ops.remove(block);
        }
      } else if (newBlock) {
        children[i] = newBlock;
        newBlock.ops.mountBefore(newBlock, anchors[i]);
      }
    }
  },
  moveBefore(block: any, anchor: any) {
    const children = block.content;
    const anchors = block.data!;
    for (let i = 0, l = children.length; i < l; i++) {
      let child = children[i];
      let _anchor = anchors[i];
      anchor.before(_anchor);
      if (child) {
        child.ops.moveBefore(child, _anchor);
      }
    }
  },
  remove(block: any) {
    const children = block.content;
    const anchors = block.data;
    for (let i = 0; i < children.length; i++) {
      const child = children[i];
      if (child) {
        child.ops.remove(child);
      }
      anchors![i].remove();
    }
  },
  firstChildNode(block: any): ChildNode | null {
    const children = block.content;
    for (let i = 0, l = children.length; i < l; i++) {
      const child = children[i];
      if (child) {
        return child.el || child.ops.firstChildNode(child);
      }
    }
    return null;
  },
};
