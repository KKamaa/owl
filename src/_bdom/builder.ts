import { elem } from "./blockdom";
import { Builder } from "./types";

// -----------------------------------------------------------------------------
//  makeBuilder
// -----------------------------------------------------------------------------
interface BuilderContext {
  path: string[];
  isDynamic: boolean;
  hasChild: boolean;
  hasHandler: boolean;
  info: {
    index: number;
    type: "text" | "child" | "handler" | "attribute";
    path: string;
    event?: string;
    name?: string;
  }[];
}

function toDom(node: ChildNode, ctx: BuilderContext): HTMLElement | Text | Comment {
  switch (node.nodeType) {
    case 1: {
      // HTMLElement
      const tagName = (node as Element).tagName;
      if (tagName.startsWith("owl-text-")) {
        const index = parseInt(tagName.slice(9), 10);
        ctx.info.push({ index, path: ctx.path.join("."), type: "text" });
        ctx.isDynamic = true;
        return document.createTextNode("");
      }
      if (tagName.startsWith("owl-child-")) {
        const index = parseInt(tagName.slice(10), 10);
        ctx.info.push({ index, type: "child", path: ctx.path.join(".") });
        ctx.hasChild = true;
        return document.createTextNode("");
      }
      const result = document.createElement((node as Element).tagName);
      const attrs = (node as Element).attributes;
      for (let i = 0; i < attrs.length; i++) {
        const attrName = attrs[i].name;
        const attrValue = attrs[i].value;
        if (attrName.startsWith("owl-handler-")) {
          const index = parseInt(attrName.slice(12), 10);
          ctx.info.push({ index, path: ctx.path.join("."), type: "handler", event: attrValue });
          ctx.hasHandler = true;
        } else if (attrName.startsWith("owl-attribute-")) {
          const index = parseInt(attrName.slice(14), 10);
          ctx.info.push({ index, path: ctx.path.join("."), type: "attribute", name: attrValue });
          ctx.isDynamic = true;
        } else {
          result.setAttribute(attrs[i].name, attrValue);
        }
      }
      let children = (node as Element).childNodes;
      const initialPath = ctx.path.slice();
      let currentPath = initialPath.slice();
      for (let i = 0; i < children.length; i++) {
        currentPath = currentPath.concat(i === 0 ? "firstChild" : "nextSibling");
        ctx.path = currentPath;
        result.appendChild(toDom(children[i], ctx));
      }
      ctx.path = initialPath;
      return result;
    }
    case 3: {
      // text node
      return document.createTextNode(node.textContent!);
    }
    case 8: {
      // comment node
      return document.createComment(node.textContent!);
    }
  }
  throw new Error("boom");
}

interface CompiledOutput {
  template: HTMLElement;
  fn: Function;
}
export function _compileBlock(str: string): CompiledOutput {
  const info: BuilderContext["info"] = [];
  const ctx: BuilderContext = {
    path: ["el"],
    info,
    hasChild: false,
    hasHandler: false,
    isDynamic: false,
  };

  const doc = new DOMParser().parseFromString(str, "text/xml");
  const template = toDom(doc.firstChild!, ctx) as any;
  const isDynamic = ctx.isDynamic;
  const hasChild = ctx.hasChild;
  const hasHandler = ctx.hasHandler;

  // console.log(context)
  const signature = hasChild ? "data, children" : isDynamic || hasHandler ? "data" : "";
  const code: string[] = [
    `  return class Builder {`,
    `    constructor(${signature}) {`,
    `      this.el = template.cloneNode(true);`,
  ];

  // preparing all internal refs
  if (info.length) {
    for (let i = 0; i < info.length; i++) {
      code.push(`      let ref${i} = this.${info[i].path};`);
    }
    code.push(
      `      this.refs = [${info.map((t: any, index: number) => `ref${index}`).join(", ")}];`
    );
  }

  if (isDynamic || hasHandler) {
    code.push(`      this.data = [];`);
  }
  if (hasChild) {
    code.push(`      this.children = [];`);
  }
  if (hasHandler) {
    // todo
    code.push(`      const handler = this.handleEvent;`);
    for (let i = 0; i < info.length; i++) {
      if (info[i].type === "handler") {
        code.push(
          `      ref${i}.addEventListener("${info[i].event}", handler.bind(this, ${info[i].index}));`
        );
      }
    }
  }
  if (info.length) {
    code.push(`      this.update(${signature});`);
  }

  // end of constructor
  code.push(`    }`);
  code.push(`    update(${signature}) {`);

  // update function
  if (info.length) {
    if (isDynamic || hasChild) {
      code.push(`      const refs = this.refs;`);
    }
    if (isDynamic || hasHandler) {
      code.push(`      const currentData = this.data;`);
      code.push(`      const nextData = data.data;`);
    }
    if (hasChild) {
      code.push(`      const currentChildren = this.children;`);
    }

    for (let i = 0; i < info.length; i++) {
      const data = info[i];
      const index = data.index;
      switch (data.type) {
        case "text":
          code.push(
            `      if (nextData[${index}] !== currentData[${index}]) { refs[${i}].textContent = nextData[${index}]}`
          );
          break;
        case "attribute":
          if (data.name === "class") {
            code.push(
              `      if (nextData[${index}] !== currentData[${index}]) { updateClass(refs[${i}], currentData[${index}], nextData[${index}]); }`
            );
          } else {
            code.push(
              `      if (nextData[${index}] !== currentData[${index}]) { let value = nextData[${index}]; refs[${i}].setAttribute("${data.name}", value === true ? "" : value);}`
            );
          }
          break;
        case "child":
          code.push(
            `      let currentChild${index} = currentChildren[${index}], child${index} = children[${index}];`
          );
          code.push(
            `      if (currentChild${index}) { if (child${index}) { currentChild${index}.ops.patch(currentChild${index}, child${index}); } else { currentChild${index}.ops.remove(currentChild${index}); currentChildren[${index}] = null; } }`
          );
          code.push(
            `      else if (child${index}) { child${index}.ops.mountBefore(child${index}, refs[${i}]); currentChildren[${index}] = child${index} }`
          );
      }
    }
    if (isDynamic || hasHandler) {
      code.push(`      this.data = nextData;`);
    }
  }
  code.push(`    }`);
  if (hasHandler) {
    // todo: move handleEvent outside of compiled code (like updateClass)
    code.push(`    handleEvent(n) {`);
    code.push(`      const handler = this.data[n];`);
    code.push(`      if (typeof handler === "function") {`);
    code.push(`        handler();`);
    code.push(`      } else {`);
    code.push(`        const [owner, method] = handler;`);
    code.push(`        owner[method]();`);
    code.push(`      }`);
    code.push(`    }`);
  }
  code.push(`  }`);

  //   console.warn(code.join("\n"));
  const wrapper = new Function("template, updateClass", code.join("\n"));
  return { template, fn: wrapper };
}

export function makeBlock(str: string): Builder {
  const { template, fn } = _compileBlock(str);
  const block = fn(template, updateClass);
  return elem.bind(null, block);
}

// -----------------------------------------------------------------------------
// helpers
// -----------------------------------------------------------------------------

function updateClass(elem: HTMLElement, prev: any, next: any) {
  prev = prev === undefined ? {} : toClassObj(prev);
  next = next === undefined ? {} : toClassObj(next);
  // remove classes
  for (let c in prev) {
    if (!(c in next)) {
      elem.classList.remove(c);
    }
  }
  // add classes
  for (let c in next) {
    if (!(c in prev)) {
      elem.classList.add(c);
    }
  }
}

function toClassObj(expr: string | number | { [c: string]: any }) {
  const result: { [c: string]: any } = {};

  if (typeof expr === "object") {
    // this is already an object but we may need to split keys:
    // {'a': true, 'b c': true} should become {a: true, b: true, c: true}
    for (let key in expr) {
      const value = expr[key];
      if (value) {
        const words = key.split(/\s+/);
        for (let word of words) {
          result[word] = value;
        }
      }
    }
    return result;
  }
  if (typeof expr !== "string") {
    expr = String(expr);
  }
  // we transform here a list of classes into an object:
  //  'hey you' becomes {hey: true, you: true}
  const str = expr.trim();
  if (!str) {
    return {};
  }
  let words = str.split(/\s+/);
  for (let i = 0, l = words.length; i < l; i++) {
    result[words[i]] = true;
  }
  return result;
}
