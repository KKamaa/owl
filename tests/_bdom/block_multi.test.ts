import { mount, multi, patch, remove, text } from "../../src/_bdom/blockdom";
import { makeBlock as origMakeBlock, _compileBlock } from "../../src/_bdom/builder";
import { makeTestFixture } from "../helpers";

function makeBlock(str: string) {
  const { fn } = _compileBlock(str);
  expect(fn.toString()).toMatchSnapshot();
  return origMakeBlock(str);
}

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

afterEach(() => {
  fixture.remove();
});

describe("multi blocks", () => {
  test("multiblock with 2 text blocks", async () => {
    const tree = multi([text("foo"), text("bar")]);
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("foobar");
  });

  test("a multiblock can be removed and leaves no extra text nodes", async () => {
    const block1 = makeBlock("<div>foo</div>");
    const block2 = makeBlock("<span>bar</span>");

    const tree = multi([block1(), block2()]);

    expect(fixture.childNodes.length).toBe(0);
    mount(tree, fixture);
    expect(fixture.childNodes.length).toBe(4);
    remove(tree);
    expect(fixture.childNodes.length).toBe(0);
  });

  test("multiblock with an empty children", async () => {
    const block = makeBlock("<div>foo</div>");
    const tree = multi([block(), undefined]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div>foo</div>");
  });

  test("multi block in a regular block", async () => {
    const block1 = makeBlock("<div><owl-child-0/></div>");
    const block2 = makeBlock("<span>yip yip</span>");

    const tree = block1([], [multi([block2()])]);

    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><span>yip yip</span></div>");
  });

  test("patching a multiblock ", async () => {
    const tree = multi([text("foo"), text("bar")]);
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("foobar");

    patch(tree, multi([text("blip"), text("bar")]));
    expect(fixture.innerHTML).toBe("blipbar");
  });
});
