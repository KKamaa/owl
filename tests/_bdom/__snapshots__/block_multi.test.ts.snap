// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multi blocks a multiblock can be removed and leaves no extra text nodes 1`] = `
"function anonymous(template, mountBefore, patch, remove, updateClass
) {
  return class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }
}"
`;

exports[`multi blocks a multiblock can be removed and leaves no extra text nodes 2`] = `
"function anonymous(template, mountBefore, patch, remove, updateClass
) {
  return class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }
}"
`;

exports[`multi blocks multi block in a regular block 1`] = `
"function anonymous(template, mountBefore, patch, remove, updateClass
) {
  return class Builder {
    constructor(data, children) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      this.refs = [ref0];
      this.children = [];
      this.update(data, children);
    }
    update(data, children) {
      const refs = this.refs;
      const currentChildren = this.children;
      let currentChild0 = currentChildren[0], child0 = children[0];
      if (currentChild0) { if (child0) { patch(currentChild0, child0); } else { remove(currentChild0); currentChildren[0] = null; } }
      else if (child0) { mountBefore(child0, refs[0]); currentChildren[0] = child0 }
    }
  }
}"
`;

exports[`multi blocks multi block in a regular block 2`] = `
"function anonymous(template, mountBefore, patch, remove, updateClass
) {
  return class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }
}"
`;

exports[`multi blocks multiblock with an empty children 1`] = `
"function anonymous(template, mountBefore, patch, remove, updateClass
) {
  return class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }
}"
`;
